File: parser/src/main/java/dev/mvmo/sicklang/parser/Parser.java
package dev.mvmo.sicklang.parser;

import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import dev.mvmo.sicklang.Lexer;
import dev.mvmo.sicklang.parser.ast.expression.*;
import dev.mvmo.sicklang.parser.ast.function.InfixParseFunction;
import dev.mvmo.sicklang.parser.ast.function.PrefixParseFunction;
import dev.mvmo.sicklang.parser.ast.program.ProgramNode;
import dev.mvmo.sicklang.parser.ast.statement.*;
import dev.mvmo.sicklang.parser.precedence.Precedence;
import dev.mvmo.sicklang.token.Token;
import dev.mvmo.sicklang.token.TokenType;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

import java.util.List;
import java.util.Map;

@Getter
@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
public class Parser {

    private final Lexer lexer;

    private final Map<TokenType, PrefixParseFunction> prefixParseFunctionMap;
    private final Map<TokenType, InfixParseFunction> infixParseFunctionMap;

    private final List<String> errorMessages;

    private Token currentToken;
    private Token peekToken;

    public static Parser newInstance(Lexer lexer) {
        var parser = new Parser(lexer, Maps.newHashMap(), Maps.newHashMap(), Lists.newArrayList());

        parser.prefixParseFunctionMap.put(TokenType.IDENTIFIER, parser::parseIdentifier);
        parser.prefixParseFunctionMap.put(TokenType.INTEGER, parser::parseIntegerLiteral);
        parser.prefixParseFunctionMap.put(TokenType.BANG, parser::parsePrefixExpression);
        parser.prefixParseFunctionMap.put(TokenType.MINUS, parser::parsePrefixExpression);
        parser.prefixParseFunctionMap.put(TokenType.TRUE, parser::parseBooleanExpression);
        parser.prefixParseFunctionMap.put(TokenType.FALSE, parser::parseBooleanExpression);
        parser.prefixParseFunctionMap.put(TokenType.LEFT_PAREN, parser::parseGroupedExpression);
        parser.prefixParseFunctionMap.put(TokenType.IF, parser::parseIfExpression);
        parser.prefixParseFunctionMap.put(TokenType.FUNCTION, parser::parseFunctionExpression);
        parser.prefixParseFunctionMap.put(TokenType.STRING, parser::parseStringLiteralExpression);
        parser.prefixParseFunctionMap.put(TokenType.LEFT_BRACKET, parser::parseArrayLiteralExpression);
        parser.prefixParseFunctionMap.put(TokenType.LEFT_BRACE, parser::parseHashLiteral);

        parser.infixParseFunctionMap.put(TokenType.PLUS, parser::parseInfixExpression);
        parser.infixParseFunctionMap.put(TokenType.MINUS, parser::parseInfixExpression);
        parser.infixParseFunctionMap.put(TokenType.SLASH, parser::parseInfixExpression);
        parser.infixParseFunctionMap.put(TokenType.ASTERISK, parser::parseInfixExpression);
        parser.infixParseFunctionMap.put(TokenType.EQUALS, parser::parseInfixExpression);
        parser.infixParseFunctionMap.put(TokenType.NOT_EQUALS, parser::parseInfixExpression);
        parser.infixParseFunctionMap.put(TokenType.LESS_THAN, parser::parseInfixExpression);
        parser.infixParseFunctionMap.put(TokenType.GREATER_THAN, parser::parseInfixExpression);
        parser.infixParseFunctionMap.put(TokenType.LEFT_PAREN, parser::parseCallExpression);
        parser.infixParseFunctionMap.put(TokenType.LEFT_BRACKET, parser::parseIndexExpression);

        parser.nextToken();
        parser.nextToken();

        return parser;
    }

    public void nextToken() {
        currentToken = peekToken;
        peekToken = lexer.nextToken();
    }

    public ProgramNode parseProgram() {
        var programNode = ProgramNode.newInstance();

        while (!currentTokenIs(TokenType.EOF)) {
            var statementNode = parseStatement();
            if (statementNode != null)
                programNode.statementNodes().add(statementNode);

            nextToken();
        }

        return programNode;
    }

    private StatementNode parseStatement() {
        return switch (currentToken.type()) {
            case LET -> parseLetStatement();
            case RETURN -> parseReturnStatement();
            default -> parseExpressionStatement();
        };
    }

    public LetStatementNode parseLetStatement() {
        var statementNode = new LetStatementNode(currentToken);
        if (!expectPeek(TokenType.IDENTIFIER))
            return null;

        statementNode.setIdentifier(IdentifierExpressionNode.newInstance(currentToken, currentToken.literal()));

        if (!expectPeek(TokenType.ASSIGN))
            return null;

        nextToken();

        statementNode.setValue(parseExpression(Precedence.LOWEST));

        if (peekTokenIs(TokenType.SEMICOLON))
            nextToken();

        return statementNode;
    }

    public ReturnStatementNode parseReturnStatement() {
        var statementNode = new ReturnStatementNode(currentToken);
        nextToken();

        statementNode.setReturnValue(parseExpression(Precedence.LOWEST));

        if (peekTokenIs(TokenType.SEMICOLON))
            nextToken();

        return statementNode;
    }

    public ExpressionStatementNode parseExpressionStatement() {
        var statementNode = ExpressionStatementNode.newInstance(currentToken);

        statementNode.expressionNode(parseExpression(Precedence.LOWEST));

        if (peekTokenIs(TokenType.SEMICOLON))
            nextToken();

        return statementNode;
    }

    public ExpressionNode parseExpression(Precedence precedence) {
        var prefixParseFunction = prefixParseFunctionMap.get(currentToken.type());
        if (prefixParseFunction == null) {
            noPrefixParseFunctionError(currentToken.type());
            return null;
        }

        var leftExpression = prefixParseFunction.parse();

        while (!peekTokenIs(TokenType.SEMICOLON) && precedence.ordinal() < Precedence.findPrecedence(peekToken.type()).ordinal()) {
            var infixParseFunction = infixParseFunctionMap.get(peekToken.type());
            if (infixParseFunction == null) {
                return leftExpression;
            }

            nextToken();

            leftExpression = infixParseFunction.parse(leftExpression);
        }

        return leftExpression;
    }

    public IdentifierExpressionNode parseIdentifier() {
        return IdentifierExpressionNode.newInstance(currentToken, currentToken.literal());
    }

    public IntegerLiteralExpressionNode parseIntegerLiteral() {
        var literalExpressionNode = IntegerLiteralExpressionNode.newInstance(currentToken);

        try {
            int integer = Integer.parseInt(currentToken.literal());
            literalExpressionNode.value(integer);
        } catch (NumberFormatException exception) {
            var errorMsg = String.format("could not parse %s as an integer", currentToken.literal());
            errorMessages.add(errorMsg);

            return null;
        }

        return literalExpressionNode;
    }

    public ExpressionNode parsePrefixExpression() {
        var prefixExpressionNode = PrefixExpressionNode.newInstance(currentToken, currentToken.literal());

        nextToken();

        prefixExpressionNode.right(parseExpression(Precedence.PREFIX));

        return prefixExpressionNode;
    }

    public ExpressionNode parseInfixExpression(ExpressionNode left) {
        var infixExpressionNode = InfixExpressionNode.newInstance(currentToken, left, currentToken.literal());
        var precedence = Precedence.findPrecedence(currentToken.type());

        nextToken();

        infixExpressionNode.right(parseExpression(precedence));

        return infixExpressionNode;
    }

    public BooleanExpressionNode parseBooleanExpression() {
        return BooleanExpressionNode.newInstance(currentToken, currentTokenIs(TokenType.TRUE));
    }

    public ExpressionNode parseGroupedExpression() {
        nextToken();

        var expressionNode = parseExpression(Precedence.LOWEST);

        if (!expectPeek(TokenType.RIGHT_PAREN))
            return null;

        return expressionNode;
    }

    public ExpressionNode parseIfExpression() {
        var expressionNode = IfExpressionNode.newInstance(currentToken);

        if (!expectPeek(TokenType.LEFT_PAREN))
            return null;

        nextToken();

        expressionNode.conditionalExpressionNode(parseExpression(Precedence.LOWEST));

        if (!expectPeek(TokenType.RIGHT_PAREN))
            return null;

        BlockStatementNode consequenceNode;
        nextToken();
        if (!currentTokenIs(TokenType.LEFT_BRACE))
            consequenceNode = parseSingleLineBlockStatement();
        else
            consequenceNode = parseBlockStatement();

        if (consequenceNode == null)
            return null;

        expressionNode.consequence(consequenceNode);

        if (peekTokenIs(TokenType.ELSE)) {
            nextToken();

            BlockStatementNode alternativeNode;
            nextToken();
            if (!currentTokenIs(TokenType.LEFT_BRACE))
                alternativeNode = parseSingleLineBlockStatement();
            else
                alternativeNode = parseBlockStatement();

            if (alternativeNode == null)
                return null;

            expressionNode.alternative(alternativeNode);
        }

        return expressionNode;
    }

    public BlockStatementNode parseSingleLineBlockStatement() {
        var blockStatementNode = BlockStatementNode.newInstance(new Token(TokenType.LEFT_BRACE, "{"));

        var statement = parseStatement();
        if (statement != null)
            blockStatementNode.statementNodes().add(statement);
        else
            return null;

        return blockStatementNode;
    }

    public BlockStatementNode parseBlockStatement() {
        var blockStatementNode = BlockStatementNode.newInstance(currentToken);

        nextToken();

        while (!currentTokenIs(TokenType.RIGHT_BRACE) && !currentTokenIs(TokenType.EOF)) {
            var statement = parseStatement();
            if (statement != null)
                blockStatementNode.statementNodes().add(statement);

            nextToken();
        }

        return blockStatementNode;
    }

    public FunctionLiteralExpressionNode parseFunctionExpression() {
        var functionLiteralExpressionNode = FunctionLiteralExpressionNode.newInstance(currentToken);

        if (!expectPeek(TokenType.LEFT_PAREN))
            return null;

        functionLiteralExpressionNode.parameters(parseFunctionParameters());

        if (!expectPeek(TokenType.LEFT_BRACE))
            return null;

        functionLiteralExpressionNode.body(parseBlockStatement());

        return functionLiteralExpressionNode;
    }

    public List<IdentifierExpressionNode> parseFunctionParameters() {
        List<IdentifierExpressionNode> identifiers = Lists.newArrayList();

        if (peekTokenIs(TokenType.RIGHT_PAREN)) {
            nextToken();
            return identifiers;
        }

        nextToken();

        identifiers.add(IdentifierExpressionNode.newInstance(currentToken, currentToken.literal()));
        // use do while
        while (peekTokenIs(TokenType.COMMA)) {
            nextToken();
            nextToken();

            identifiers.add(IdentifierExpressionNode.newInstance(currentToken, currentToken.literal()));
        }

        if (!expectPeek(TokenType.RIGHT_PAREN))
            return null;

        return identifiers;
    }

    public ExpressionNode parseCallExpression(ExpressionNode function) {
        var callExpressionNode = CallExpressionNode.newInstance(currentToken, function);
        callExpressionNode.arguments(parseExpressionList(TokenType.RIGHT_PAREN));

        return callExpressionNode;
    }

    // TODO: private access
    public StringLiteralExpressionNode parseStringLiteralExpression() {
        return StringLiteralExpressionNode.newInstance(currentToken, currentToken.literal());
    }

    public ArrayLiteralExpressionNode parseArrayLiteralExpression() {
        return ArrayLiteralExpressionNode.newInstance(currentToken, parseExpressionList(TokenType.RIGHT_BRACKET));
    }

    public IndexExpressionNode parseIndexExpression(ExpressionNode left) {
        var startToken = currentToken;
        nextToken();
        var indexExpression = parseExpression(Precedence.LOWEST);

        if (!expectPeek(TokenType.RIGHT_BRACKET))
            return null;

        return IndexExpressionNode.newInstance(startToken, left, indexExpression);
    }

    public HashLiteralExpressionNode parseHashLiteral() {
        var startToken = currentToken;
        Map<ExpressionNode, ExpressionNode> pairs = Maps.newHashMap();

        while (!peekTokenIs(TokenType.RIGHT_BRACE)) {
            nextToken();
            var keyExpression = parseExpression(Precedence.LOWEST);

            if (!expectPeek(TokenType.COLON))
                return null;

            nextToken();

            var valueExpression = parseExpression(Precedence.LOWEST);
            pairs.put(keyExpression, valueExpression);

            if (!peekTokenIs(TokenType.RIGHT_BRACE) && !expectPeek(TokenType.COMMA))
                return null;
        }

        if (!expectPeek(TokenType.RIGHT_BRACE))
            return null;

        return HashLiteralExpressionNode.newInstance(startToken, pairs);
    }

    public List<ExpressionNode> parseExpressionList(TokenType till) {
        List<ExpressionNode> list = Lists.newArrayList();

        if (peekTokenIs(till)) {
            nextToken();
            return list;
        }

        nextToken();
        list.add(parseExpression(Precedence.LOWEST)); // TODO DO-WHILE?

        while (peekTokenIs(TokenType.COMMA)) {
            nextToken();
            nextToken();

            list.add(parseExpression(Precedence.LOWEST));
        }

        if (!expectPeek(till))
            return null;

        return list;
    }

    private boolean currentTokenIs(TokenType tokenType) {
        return currentToken.type().equals(tokenType);
    }

    private boolean peekTokenIs(TokenType tokenType) {
        return peekToken.type().equals(tokenType);
    }

    private boolean expectPeek(TokenType tokenType) {
        if (peekTokenIs(tokenType)) {
            nextToken();
            return true;
        } else {
            peekError(tokenType);
            return false;
        }
    }

    private void noPrefixParseFunctionError(TokenType type) {
        String message = String.format("no prefix parse function found for type %s", type.name());
        errorMessages.add(message);
    }

    private void peekError(TokenType tokenType) {
        String message = String.format("expected next token to be %s, got %s instead", tokenType, peekToken.type());
        errorMessages.add(message);
    }

}

File: parser/src/main/java/dev/mvmo/sicklang/parser/Parser.kt
package dev.mvmo.sicklang.parser;

import dev.mvmo.sicklang.Lexer
import dev.mvmo.sicklang.parser.ast.expression.*
import dev.mvmo.sicklang.parser.ast.function.InfixParseFunction
import dev.mvmo.sicklang.parser.ast.function.PrefixParseFunction
import dev.mvmo.sicklang.parser.ast.program.ProgramNode
import dev.mvmo.sicklang.parser.ast.statement.*
import dev.mvmo.sicklang.parser.precedence.Precedence
import dev.mvmo.sicklang.token.Token
import dev.mvmo.sicklang.token.TokenType

class Parser(private val lexer: Lexer) {

    private val prefixParseFunctionMap = mutableMapOf<TokenType, PrefixParseFunction>()
    private val infixParseFunctionMap = mutableMapOf<TokenType, InfixParseFunction>()

    val errorMessages: MutableList<String> = arrayListOf()

    private var currentToken: Token
    private var peekToken: Token

    init {
        with(prefixParseFunctionMap) {
            put(TokenType.IDENTIFIER, PrefixParseFunction { parseIdentifier() })
            put(TokenType.INTEGER, PrefixParseFunction { parseIntegerLiteral() })
            put(TokenType.BANG, PrefixParseFunction { parsePrefixExpression() })
            put(TokenType.MINUS, PrefixParseFunction { parsePrefixExpression() })
            put(TokenType.TRUE, PrefixParseFunction { parseBooleanExpression() })
            put(TokenType.FALSE, PrefixParseFunction { parseBooleanExpression() })
            put(TokenType.LEFT_PAREN, PrefixParseFunction { parseGroupedExpression() })
            put(TokenType.IF, PrefixParseFunction { parseIfExpression() })
            put(TokenType.FUNCTION, PrefixParseFunction { parseFunctionExpression() })
            put(TokenType.STRING, PrefixParseFunction { parseStringLiteralExpression() })
            put(TokenType.LEFT_BRACKET, PrefixParseFunction { parseArrayLiteralExpression() })
            put(TokenType.LEFT_BRACE, PrefixParseFunction { parseHashLiteral() })
        }

        with(infixParseFunctionMap) {
            put(TokenType.PLUS, InfixParseFunction { parseInfixExpression(it) })
            put(TokenType.MINUS, InfixParseFunction { parseInfixExpression(it) })
            put(TokenType.SLASH, InfixParseFunction { parseInfixExpression(it) })
            put(TokenType.ASTERISK, InfixParseFunction { parseInfixExpression(it) })
            put(TokenType.EQUALS, InfixParseFunction { parseInfixExpression(it) })
            put(TokenType.NOT_EQUALS, InfixParseFunction { parseInfixExpression(it) })
            put(TokenType.LESS_THAN, InfixParseFunction { parseInfixExpression(it) })
            put(TokenType.GREATER_THAN, InfixParseFunction { parseInfixExpression(it) })
            put(TokenType.LEFT_PAREN, InfixParseFunction { parseCallExpression(it) })
            put(TokenType.LEFT_BRACKET, InfixParseFunction { parseIndexExpression(it) })
            put(TokenType.OR, InfixParseFunction { parseInfixExpression(it) })
            put(TokenType.AND, InfixParseFunction { parseInfixExpression(it) })
            put(TokenType.BITWISE_AND, InfixParseFunction { parseInfixExpression(it) })
            put(TokenType.BITWISE_OR, InfixParseFunction { parseInfixExpression(it) })
        }

        this.currentToken = lexer.nextToken()
        this.peekToken = currentToken

        nextToken()
    }

    fun nextToken() {
        this.currentToken = peekToken
        this.peekToken = lexer.nextToken()
    }

    fun parseProgram(): ProgramNode {
        val programNode = ProgramNode()

        while (!currentTokenIs(TokenType.EOF)) {
            val statementNode = parseStatement()
            programNode.statementNodes.add(statementNode);

            nextToken();
        }

        return programNode;
    }

    fun parseStatement(): StatementNode {
        return when (currentToken.type) {
            TokenType.LET -> parseLetStatement()
            TokenType.RETURN -> parseReturnStatement()
            else -> parseExpressionStatement()
        }
    }

    fun parseLetStatement(): LetStatementNode {
        val statementNode = LetStatementNode(currentToken)
        expectPeek(TokenType.IDENTIFIER)

        statementNode.identifier = IdentifierExpressionNode(currentToken, currentToken.literal)

        expectPeek(TokenType.ASSIGN)

        nextToken()

        statementNode.value = parseExpression(Precedence.LOWEST)

        if (peekTokenIs(TokenType.SEMICOLON))
            nextToken()

        return statementNode;
    }

    fun parseReturnStatement(): ReturnStatementNode {
        val statementNode = ReturnStatementNode(currentToken)
        nextToken()

        statementNode.returnValue = parseExpression(Precedence.LOWEST)

        if (peekTokenIs(TokenType.SEMICOLON))
            nextToken()

        return statementNode;
    }

    fun parseExpressionStatement(): ExpressionStatementNode {
        val statementNode = ExpressionStatementNode(currentToken)

        statementNode.expressionNode = parseExpression(Precedence.LOWEST)

        if (peekTokenIs(TokenType.SEMICOLON))
            nextToken()

        return statementNode
    }

    fun parseExpression(precedence: Precedence): ExpressionNode {
        val prefixParseFunction = prefixParseFunctionMap[currentToken.type]
            ?: throw SickParseException("No prefix parse function found for ${currentToken.type}")

        var leftExpression = prefixParseFunction.parse()

        while (!peekTokenIs(TokenType.SEMICOLON) && precedence.ordinal < Precedence.findPrecedence(peekToken.type).ordinal) {
            val infixParseFunction = infixParseFunctionMap[peekToken.type]
                ?: return leftExpression

            nextToken();

            leftExpression = infixParseFunction.parse(leftExpression)
        }

        return leftExpression;
    }

    fun parseIdentifier(): IdentifierExpressionNode {
        return IdentifierExpressionNode(currentToken, currentToken.literal);
    }

    fun parseIntegerLiteral(): IntegerLiteralExpressionNode {
        val literalExpressionNode = IntegerLiteralExpressionNode(currentToken)

        try {
            val integer = currentToken.literal.toInt()
            literalExpressionNode.value = integer
        } catch (_: NumberFormatException) {
            throw SickParseException("Couldn't convert '${currentToken.literal}' to int")
        }

        return literalExpressionNode
    }

    fun parsePrefixExpression(): ExpressionNode {
        val prefixExpressionNode = PrefixExpressionNode(currentToken, currentToken.literal)

        nextToken()
        prefixExpressionNode.right = parseExpression(Precedence.PREFIX)

        return prefixExpressionNode
    }

    fun parseInfixExpression(left: ExpressionNode): ExpressionNode {
        val infixExpressionNode = InfixExpressionNode(currentToken, left, currentToken.literal)
        val precedence = Precedence.findPrecedence(currentToken.type)

        nextToken()

        infixExpressionNode.right = parseExpression(precedence)

        return infixExpressionNode
    }

    fun parseBooleanExpression(): BooleanExpressionNode {
        return BooleanExpressionNode(currentToken, currentTokenIs(TokenType.TRUE))
    }

    fun parseGroupedExpression(): ExpressionNode {
        nextToken()

        val expressionNode = parseExpression(Precedence.LOWEST)
        expectPeek(TokenType.RIGHT_PAREN)

        return expressionNode
    }

    fun parseIfExpression(): ExpressionNode {
        val expressionNode = IfExpressionNode(currentToken)

        expectPeek(TokenType.LEFT_PAREN)

        nextToken()

        expressionNode.conditionalExpressionNode = parseExpression(Precedence.LOWEST)

        expectPeek(TokenType.RIGHT_PAREN)

        nextToken()
        val consequenceNode = if (!currentTokenIs(TokenType.LEFT_BRACE))
            parseSingleLineBlockStatement()
        else
            parseBlockStatement()

        expressionNode.consequence = consequenceNode

        if (peekTokenIs(TokenType.ELSE)) {
            nextToken()
            nextToken()

            val alternativeNode = if (!currentTokenIs(TokenType.LEFT_BRACE))
                parseSingleLineBlockStatement()
            else
                parseBlockStatement()

            expressionNode.alternative = alternativeNode
        }

        return expressionNode;
    }

    fun parseSingleLineBlockStatement(): BlockStatementNode {
        val blockStatementNode = BlockStatementNode(Token(TokenType.LEFT_BRACE, "{"))

        val statement = parseStatement()
        blockStatementNode.statementNodes.add(statement)

        return blockStatementNode;
    }

    fun parseBlockStatement(): BlockStatementNode {
        val blockStatementNode = BlockStatementNode(currentToken)

        nextToken()

        while (!currentTokenIs(TokenType.RIGHT_BRACE) && !currentTokenIs(TokenType.EOF)) {
            val statement = parseStatement()
            blockStatementNode.statementNodes.add(statement)

            nextToken()
        }

        return blockStatementNode
    }

    fun parseFunctionExpression(): FunctionLiteralExpressionNode {
        val functionLiteralExpressionNode = FunctionLiteralExpressionNode(currentToken)

        expectPeek(TokenType.LEFT_PAREN)

        functionLiteralExpressionNode.parameters = parseFunctionParameters()

        expectPeek(TokenType.LEFT_BRACE)

        functionLiteralExpressionNode.body = parseBlockStatement()

        return functionLiteralExpressionNode
    }

    fun parseFunctionParameters(): List<IdentifierExpressionNode> {
        val identifiers = arrayListOf<IdentifierExpressionNode>()

        if (peekTokenIs(TokenType.RIGHT_PAREN)) {
            nextToken()
            return identifiers
        }

        nextToken()

        identifiers.add(IdentifierExpressionNode(currentToken, currentToken.literal))

        while (peekTokenIs(TokenType.COMMA)) {
            nextToken()
            nextToken()

            identifiers.add(IdentifierExpressionNode(currentToken, currentToken.literal))
        }

        expectPeek(TokenType.RIGHT_PAREN)

        return identifiers
    }

    fun parseCallExpression(function: ExpressionNode): ExpressionNode {
        val callExpressionNode = CallExpressionNode(currentToken, function)
        callExpressionNode.arguments = parseExpressionList(TokenType.RIGHT_PAREN)

        return callExpressionNode
    }

    fun parseStringLiteralExpression(): StringLiteralExpressionNode {
        return StringLiteralExpressionNode(currentToken, currentToken.literal)
    }

    fun parseArrayLiteralExpression(): ArrayLiteralExpressionNode {
        return ArrayLiteralExpressionNode(currentToken, parseExpressionList(TokenType.RIGHT_BRACKET))
    }

    fun parseIndexExpression(left: ExpressionNode): IndexExpressionNode {
        val startToken = currentToken
        nextToken()
        val indexExpression = parseExpression(Precedence.LOWEST)

        expectPeek(TokenType.RIGHT_BRACKET)

        return IndexExpressionNode(startToken, left, indexExpression)
    }

    private fun parseHashLiteral(): HashLiteralExpressionNode {
        val startToken = currentToken
        val pairs = mutableMapOf<ExpressionNode, ExpressionNode>()

        while (!peekTokenIs(TokenType.RIGHT_BRACE)) {
            nextToken();
            val keyExpression = parseExpression(Precedence.LOWEST);

            expectPeek(TokenType.COLON)

            nextToken()

            val valueExpression = parseExpression(Precedence.LOWEST)
            pairs[keyExpression] = valueExpression

            if (!peekTokenIs(TokenType.RIGHT_BRACE))
                expectPeek(TokenType.COMMA)
        }

        expectPeek(TokenType.RIGHT_BRACE)

        return HashLiteralExpressionNode(startToken, pairs)
    }

    private fun parseExpressionList(until: TokenType): List<ExpressionNode> {
        val list = arrayListOf<ExpressionNode>()

        if (peekTokenIs(until)) {
            nextToken()
            return list
        }

        nextToken()
        list.add(parseExpression(Precedence.LOWEST))

        while (peekTokenIs(TokenType.COMMA)) {
            nextToken()
            nextToken()

            list.add(parseExpression(Precedence.LOWEST))
        }

        expectPeek(until)

        return list
    }

    fun currentTokenIs(tokenType: TokenType): Boolean {
        return currentToken.type == tokenType
    }

    fun peekTokenIs(tokenType: TokenType): Boolean {
        return peekToken.type == tokenType
    }

    private fun expectPeek(tokenType: TokenType) {
        if (peekTokenIs(tokenType)) {
            nextToken()
            return
        }

        throw UnexpectedTokenException(tokenType, peekToken.type)
    }

}
